controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/RevisionHistory.java

```diff
@@ -0,0 +1,147 @@
+
+    private RevisionHistory(NavigableMap<RevisionId, ApplicationVersion> production,
+                           NavigableMap<JobId, NavigableMap<RevisionId, ApplicationVersion>> development) {
+        this.production = production;
+        this.development = development;
+    }
+
+    public static RevisionHistory empty() {
+        return ofRevisions(List.of(), Map.of());
+    }
+
+    public static RevisionHistory ofRevisions(Collection<ApplicationVersion> productionRevisions,
+                                              Map<JobId, ? extends Collection<ApplicationVersion>> developmentRevisions) {
+        NavigableMap<RevisionId, ApplicationVersion> production = new TreeMap<>();
+        for (ApplicationVersion revision : productionRevisions)
+            production.put(revision.id(), revision);
+
+        NavigableMap<JobId, NavigableMap<RevisionId, ApplicationVersion>> development = new TreeMap<>(comparator);
+        developmentRevisions.forEach((job, jobRevisions) -> {
+            NavigableMap<RevisionId, ApplicationVersion> revisions = development.computeIfAbsent(job, __ -> new TreeMap<>());
+            for (ApplicationVersion revision : jobRevisions)
+                revisions.put(revision.id(), revision);
+        });
+
+        return new RevisionHistory(production, development);
+    }
+
+    /** Returns a copy of this with given production revision forgotten. */
+    public RevisionHistory without(RevisionId id) {
+        if ( ! production.containsKey(id)) return this;
+        TreeMap<RevisionId, ApplicationVersion> production = new TreeMap<>(this.production);
+        production.remove(id);
+        return new RevisionHistory(production, development);
+    }
+
+    /** Returns a copy of this with the given development revision forgotten. */
+    public RevisionHistory without(RevisionId id, JobId job) {
+        if ( ! development.containsKey(job) || ! development.get(job).containsKey(id)) return this;
+        NavigableMap<JobId, NavigableMap<RevisionId, ApplicationVersion>> development = new TreeMap<>(this.development);
+        development.get(job).remove(id);
+        return new RevisionHistory(production, development);
+    }
+
+    /** Returns a copy of this with the production revision added or updated */
+    public RevisionHistory with(ApplicationVersion revision) {
+        NavigableMap<RevisionId, ApplicationVersion> production = new TreeMap<>(this.production);
+        production.put(revision.id(), revision);
+        return new RevisionHistory(production, development);
+    }
+
+    /** Returns a copy of this with the new development revision added, and the previous version without a package. */
+    public RevisionHistory with(ApplicationVersion revision, JobId job) {
+        NavigableMap<JobId, NavigableMap<RevisionId, ApplicationVersion>> development = new TreeMap<>(this.development);
+        NavigableMap<RevisionId, ApplicationVersion> revisions = development.computeIfAbsent(job, __ -> new TreeMap<>());
```

Comment:
By returning existing here we will modify the original instance

Reply:
Yes, thanks, forgot a copy here!